packages/jyve-pyodide-unsafe/src/kernel.ts

```diff
@@ -0,0 +1,121 @@
+import {Kernel, KernelMessage} from '@jupyterlab/services';
+
+import {JSUnsafeKernel} from '@deathbeds/jyve-js-unsafe';
+import {JyveKernel} from '@deathbeds/jyve/lib/kernel';
+
+const {jyve} = (require('../package.json') as any);
+
+export const kernelSpec: Kernel.ISpecModel = jyve.kernelspec;
+
+// const DEBUG = false;
+
+
+export class PyodideUnsafeKernel extends JSUnsafeKernel {
+  protected kernelSpec = kernelSpec;
+  protected _pyodide: any;
+
+  jyveInfo(): KernelMessage.IInfoReply {
+    const jsInfo = super.jyveInfo();
+    return {
+      ...jsInfo,
+      help_links: [...jsInfo.help_links, ...jyve.help_links],
+      implementation: kernelSpec.name,
+      language_info: jyve.language_info
+    };
+  }
+
+  async pyodide() {
+    if (!this._pyodide) {
+      this._pyodide = await PyodideUnsafeKernel.pyodide(
+        (await this.iframe()).contentWindow
+      );
+    }
+    return this._pyodide;
+  }
+
+  resetUserNS() {
+    // TODO: Clear the namespace on the Python side
+    this.userNS = {};
+    this._pyodide = null;
+  }
+
+  async execNS(parent: KernelMessage.IMessage) {
+    const k = this;
+
+    let execNS = await super.execNS(parent);
+    const pyodide = await this.pyodide();
+    // pyodide.$options = {debug: DEBUG ? 10 : 0};
+
+    execNS = {
+      ...execNS,
+      __PYODIDE__: pyodide,
+    };
+
+    pyodide.write = (data: any) => {
+        k.sendJSON(k.fakeDisplayData(parent, {
+            'text/plain': `${data}`}));
+    };
+
+    pyodide.runPython(
+        "from js import window as _window\n" +
+        "import sys\n" +
+        "sys.stdout.write = _window.pyodide.write\n" +
+        "sys.stderr.write = _window.pyodide.write\n");
+
+    return execNS;
+  }
+
+  async transpile(code: string) {
+    await this.pyodide();
+    const src = JSON.stringify(code);
+
+    return `__PYODIDE__.runPython(${src});`;
+  }
+}
+
+function bootstrap_pyodide(window: any) {
+  let baseURL = "http://iodide-project.github.io/pyodide-demo/";
+  let wasmURL = `${baseURL}pyodide.asm.wasm?x=${Date.now()}`;
+  let wasmXHR = new XMLHttpRequest();
+  wasmXHR.open('GET', wasmURL, true);
+  wasmXHR.responseType = 'arraybuffer';
+  wasmXHR.onload = function() {
+    let Module: any = {};
```

Comment:
So, we'll likely see this on more kernels. I'm interested in the pattern that makes this possible reliable and non-hangy. Excited to watch what develops!

Reply:
Yes -- it definitely can / does hang if anything goes wrong.  I think as easy win is just capping a timeout on things so at least we get control back if things fall off a cliff.  Other than that, it's not clear.  I know emscripten is growing some more use of Promises that might help clean things up on this side, but it's not quite ready as far as I can tell.